staging/src/k8s.io/client-go/pkg/apis/clientauthentication/types.go

```diff
@@ -75,3 +80,42 @@ type Response struct {
 	// Code is the HTTP status code returned by the server.
 	Code int32
 }
+
+// Cluster contains information to allow an exec plugin to communicate
+// with the kubernetes cluster being authenticated to.
+type Cluster struct {
+	// Server is the address of the kubernetes cluster (https://hostname:port).
+	Server string
+	// ServerName is passed to the server for SNI and is used in the client to check server
+	// certificates against. If ServerName is empty, the hostname used to contact the
+	// server is used.
+	// +optional
+	ServerName string
+	// CAData contains PEM-encoded certificate authority certificates.
+	// If empty, system roots should be used.
+	// +listType=atomic
+	// +optional
+	CAData []byte
+	// Config holds additional config data that is specific to the exec
+	// plugin with regards to the cluster being authenticated to.
+	//
+	// This data is sourced from the clientcmd Cluster object's extensions[exec] field:
+	//
+	// clusters:
+	// - name: my-cluster
+	//   cluster:
+	//     ...
+	//     extensions:
+	//     - name: exec  # reserved extension name for per cluster exec config
```

Comment:
I'm not sure it's ok to declare an unqualified name (no `kubernetes.io/` prefix) reserved after the fact like this.

would it be better for the exec config to opt into getting cluster info (and maybe even identify the named cluster extension it wants)?

```yaml
clusters:
- name: my-cluster
  cluster:
    ...
    extensions:
    - name: myexec
      extension:
        audience: 06e3fbd18de8  # arbitrary config
users:
- name: myuser
  user:
    exec:
      ...
      includeClusterConfig: true
      includeClusterExtension: myexec
```

that would:
1. avoid having a magic extension name
2. avoid clusters only working with a single plugin
3. avoid sending exec plugins data they did not expect/want (CA data specifically could be quite large to include in the env)
4. avoid multiple calls to an exec plugin for different clusters when it didn't care about the cluster and the already-fetched credential could have been used

Reply:
I would personally lean towards just changing `exec` to `client.authentication.k8s.io/exec` to prevent any possible collision instead of expanding `ExecConfig` (though I suspect that we would be fine using `exec`).

> avoid having a magic extension name

I am not sure this is much different from the `exec:` stanza under `user:`.

> avoid clusters only working with a single plugin

Somehow I do not see a user having the same cluster with multiple distinct exec plugins configured.  I realize it is technically possible, but not exactly practical or common.

> avoid sending exec plugins data they did not expect/want (CA data specifically could be quite large to include in the env)

I think we can handle the extra 2 KB via the env var...

> avoid multiple calls to an exec plugin for different clusters when it didn't care about the cluster and the already-fetched credential could have been used

Using the same credential across clusters is an incredibly unsafe pattern.  If you are using a bearer token, you are at risk of a compromised cluster replaying your credentials.  For both tokens and certs, if the credential is stolen from disk, it can be used against multiple clusters.  Combine that with no revocation of certs, and you are really asking for trouble.  I am okay adding the latency of exec-ing the binary per cluster as the binary will likely need to do some form of caching itself to support clients like `kubectl`.

Reply:
Opened a new PR into the KEP to try to address this design update: https://github.com/kubernetes/enhancements/pull/1927.